libcaf_core/caf/flow/op/sample.hpp

```diff
@@ -0,0 +1,319 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/master/LICENSE.
+
+#pragma once
+
+#include "caf/cow_vector.hpp"
+#include "caf/flow/observable_decl.hpp"
+#include "caf/flow/observer.hpp"
+#include "caf/flow/op/cold.hpp"
+#include "caf/flow/op/state.hpp"
+#include "caf/flow/subscription.hpp"
+#include "caf/unit.hpp"
+
+#include <vector>
+
+namespace caf::flow::op {
+
+struct sample_input_t {};
+
+struct sample_emit_t {};
+
+template <class T>
+struct sample_interval_trait {
+  using input_type = T;
+  using output_type = std::optional<T>;
+  using select_token_type = int64_t;
+
+  output_type operator()(const std::optional<input_type>& xs) {
+    return output_type{xs};
+  }
+};
+
+///
+template <class Trait>
+class sample_sub : public subscription::impl_base {
```

Comment:
Looks like you've simply left the trait in here because you've based this on `buffer`. However, this trait does not configure anything now and should be dropped. Also, output type should be `T`, not `optional<T>`. If the last interval had no value, the `sample` operator simply emits nothing.

```suggestion
template <class T>
class sample_sub : public subscription::impl_base {
```